src/treelearner/serial_tree_learner.cpp

```diff
@@ -765,4 +804,51 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
   }
 }
 
+
+// this function updates the best split for a leaf
+// it is called only when monotone constraints exist
+void SerialTreeLearner::UpdateBestSplitsFromHistograms(
```

Comment:
I am confused about this function. What do you want to get in this function?
Why there are two loops over feature_index?
Do you want to update the leaf splits, by different constraints?

Reply:
Yes exactly, the only goal of this function is to update the best split (called split here), with the new constraints it just got. 

To summarize, first a split is made somewhere. Then the function GoUpToFindLeavesToUpdate updates the constraints of all other splits (if needed). Then the function UpdateBestSplitsFromHistograms updates the best splits in case the constraints associated with the previous best splits were modified. Indeed, if they were, the previous best split may not be the best split anymore.